src/Service/Bulk/BulkUserCreationService.php

```diff
@@ -0,0 +1,322 @@
+        string $generatedUsersFilePath
+    ) {
+        $this->lineItemRepository = $lineItemRepository;
+        $this->assignmentRepository = $assignmentRepository;
+        $this->passwordHasher = $passwordHasher;
+        $this->userRepository = $userRepository;
+        $this->assignmentIngester = $assignmentIngester;
+        $this->filesystem = $filesystem;
+        $this->groupIdLoadBalancerFactory = $groupIdLoadBalancerFactory;
+        $this->userCreationResponse = $userCreationResponse;
+        $this->csvWriter = $csvWriter;
+        $this->generatedUsersFilePath = $generatedUsersFilePath;
+    }
+
+    /**
+     * @throws LineItemNotFoundException
+     */
+    public function createUsers(
+        array $lineItemIds,
+        array $lineItemSlugs,
+        array $userPrefix,
+        int $batchSize,
+        ?string $groupPrefix
+    ): array {
+
+        $notExistLineItemsArray = $userGroupIds = [];
+
+        if (!empty($lineItemIds) || !empty($lineItemSlugs)) {
+            $this->lineItemIds = $lineItemIds;
+            $this->lineItemSlugs = $lineItemSlugs;
+            $criteria = $this->getFindLineItemCriteria();
+            $lineItemsArray = ($this->lineItemRepository->findLineItemsByCriteria($criteria))->jsonSerialize();
+            if (empty($lineItemsArray)) {
+                $exceptionMessage = $this->lineItemIds
+                    ? implode(',', $this->lineItemIds) . ' LineItem Ids not exist in the system'
+                    : implode(',', $this->lineItemSlugs) . ' LineItem Slugs not exist in the system';
+                throw new LineItemNotFoundException($exceptionMessage);
+            }
+
+            $lineItemSlugArray = $this->generateSlugData($lineItemsArray);
+            $notExistLineItemsArray = $this->lineItemIds
+                ? array_diff($lineItemIds, array_keys($lineItemSlugArray))
+                : array_diff($lineItemSlugs, array_values($lineItemSlugArray));
+        }
+
+        if (empty($lineItemIds) && empty($lineItemSlugs)) {
+            $this->getAllLineItemSlugs();
+        }
+
+        $userIncrNo = $this->getLastUserAssignedToLineItems($this->lineItemSlugs);
+
+        $userGroupAssignCount = 0;
+        if ($groupPrefix) {
+            $userGroupIds = $this->groupIdLoadBalancerFactory->getLoadBalanceGroupID($groupPrefix);
+            $userGroupAssignCount = (int) ceil(
+                count($userPrefix) * count($this->lineItemSlugs) * $batchSize / count($userGroupIds)
+            );
+        }
+
+        $slugTotalUsers = $this->createBulkUserAssignmentData(
+            $userIncrNo,
+            $userGroupIds,
+            $userPrefix,
+            $userGroupAssignCount,
+            $groupPrefix,
+            $batchSize
+        );
+
+        return $this->userCreationResponse->userCreationResult(
+            $slugTotalUsers,
+            $notExistLineItemsArray,
+            $userPrefix
+        );
+    }
+
+    private function generateSlugData(array $lineData): array
```

Comment:
instead of `array $lineData` I would use the `LineItemResultSet $lineItems`

And example:

```
private function generateSlugData(LineItemResultSet $lineItems): array
    {
        $lineItemSlugs = [];
        foreach ($lineItems as $lineItem) {
            $lineItemSlugs[$lineItem->getId()] = $lineItem->getSlug();
        }
        $this->lineItemSlugs = $lineItemSlugs;

        return $lineItemSlugs;
    }
  ```